File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/interfaces/NativeLoginManager.kt
/*
 * Copyright (c) 2024-present, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.salesforce.androidsdk.auth.interfaces

import android.content.Intent

enum class NativeLoginResult {
    /** Email address is not a valid format  */
    InvalidEmail,

    /** Username does not meet Salesforce criteria (length, email format, ect) */
    InvalidUsername,

    /** Password does not meet the weakest Salesforce criteria */
    InvalidPassword,

    /** Username/password combination is incorrect. */
    InvalidCredentials,

    UnknownError,

    Success,
}

/**
 *  Manage native login flow.
 */
interface NativeLoginManager {

    /**
     * If the native login view should show a back button.
     */
    val shouldShowBackButton: Boolean

    /**
     * The username of the locked account or null.  Can be used to pre-populate the username field
     * or in a message telling the user which account biometric will unlock.
     */
    val biometricAuthenticationUsername: String?

    /**
     * Initiate a login with user provided username and password.
     *
     * @param username User provided Salesforce username.
     * @param password User provided Salesforce password.
     * @return NativeLoginResult
     */
    suspend fun login(username: String, password: String): NativeLoginResult

    /**
     * Use this function to get the intent for fallback web based authentication.
     *
     * In your native login activity start this intent in such a way as to [get a result callback](https://developer.android.com/training/basics/intents/result).
     * A result code of `Activity.RESULT_OK` means the user has successfully authenticated via the webview and you
     * can finish your activity.
     *
     * @return the intent to be started for fallback web based authentication.
     */
    fun getFallbackWebAuthenticationIntent(): Intent

    // region Salesforce Identity API Headless Registration Flow

    /**
     * Submits a request to start a user registration to the Salesforce Identity
     * API headless registration flow.  This fulfills step four of the headless
     * registration flow.
     *
     * See https://help.salesforce.com/s/articleView?id=sf.remoteaccess_headless_registration_public_clients.htm&type=5
     *
     * @param email The user-entered email address
     * @param firstName The user-entered first name
     * @param lastName The user-entered last name
     * @param username A valid Salesforce username or email
     * @param newPassword The user-entered new password
     * @param reCaptchaToken A reCAPTCHA token provided by the reCAPTCHA SDK
     * @param otpVerificationMethod: The delivery method for the OTP
     * @return The start registration result with the request identifier
     * returned by the Salesforce Identity API and a native login result
     * indicating success or one of several possible failures, including both
     * in-app and Salesforce Identity API results
     */
    suspend fun startRegistration(
        email: String,
        firstName: String,
        lastName: String,
        username: String,
        newPassword: String,
        reCaptchaToken: String,
        otpVerificationMethod: OtpVerificationMethod
    ): StartRegistrationResult

    /**
     * Submits a request to complete a user registration to the Salesforce
     * Identity API headless registration flow.  This fulfills step eight of the
     * headless registration flow.
     *
     * See https://help.salesforce.com/s/articleView?id=sf.remoteaccess_headless_registration_public_clients.htm&type=5
     *
     * @param otp A user-entered one-time-password
     * @param requestIdentifier The request identifier issued by the Salesforce
     * Identity API headless registration flow in the start registration method
     * @param otpVerificationMethod The one-time-password verification method
     * used to obtain the OTP identifier
     * @return A native login result indicating success or one of several
     * possible failures, including both in-app and Salesforce Identity API
     * results
     */
    suspend fun completeRegistration(
        otp: String,
        requestIdentifier: String,
        otpVerificationMethod: OtpVerificationMethod
    ): NativeLoginResult

    /**
     * A start registration result.
     * @param nativeLoginResult A native login result indicating success or one of
     * several possible failures, including both in-app and Salesforce Identity API
     * results
     * @param email On success result, the email provided by the Salesforce Identity
     * API
     * @param requestIdentifier On success result, the request identifier provided
     * by the Salesforce Identity API
     */
    data class StartRegistrationResult(
        val nativeLoginResult: NativeLoginResult,
        val email: String? = null,
        val requestIdentifier: String? = null
    )

    // endregion
    // region Salesforce Identity API Headless Forgot Password Flow

    /**
     * Submits a request to start a password reset to the Salesforce Identity
     * API headless forgot password flow.  This fulfills step one of the
     * headless forgot password flow.
     *
     * See https://help.salesforce.com/s/articleView?id=sf.remoteaccess_headless_forgot_password_flow.htm&type=5
     *
     * @param username A valid Salesforce username or email
     * @param reCaptchaToken A reCAPTCHA token provided by the reCAPTCHA SDK
     * @return A native login result indicating success or one of several
     * possible failures, including both in-app and Salesforce Identity API
     * results
     */
    suspend fun startPasswordReset(
        username: String,
        reCaptchaToken: String
    ): NativeLoginResult

    /**
     * Submits a request to complete a password reset to the Salesforce Identity
     * API headless forgot password flow. This fulfills step four of the
     * headless forgot password flow.
     *
     * See https://help.salesforce.com/s/articleView?id=sf.remoteaccess_headless_forgot_password_flow.htm&type=5
     *
     * @param username A valid Salesforce username or email
     * @param otp A user-entered one-time-password
     * @param newPassword The user-entered new password
     * @return A native login result indicating success or one of several
     * possible failures, including both in-app and Salesforce Identity API
     * results
     */
    suspend fun completePasswordReset(
        username: String,
        otp: String,
        newPassword: String
    ): NativeLoginResult

    // endregion
    // region Salesforce Identity API Headless, Password-Less Login Via One-Time-Passcode

    /**
     * Submits a request to start password-less login via one-time-passcode to
     * the Salesforce Identity API headless, password-less login flow. This
     * fulfills step three of the headless, password-less login flow.
     *
     * See https://help.salesforce.com/s/articleView?id=sf.remoteaccess_headless_passwordless_login_public_clients.htm&type=5
     *
     * @param username A valid Salesforce username or email
     * @param reCaptchaToken A reCAPTCHA token provided by the reCAPTCHA SDK
     * @param otpVerificationMethod: The delivery method for the OTP
     * @return An OTP request result with the overall login result and the OTP
     * identifier for successful OTP requests
     */
    suspend fun submitOtpRequest(
        username: String,
        reCaptchaToken: String,
        otpVerificationMethod: OtpVerificationMethod
    ): OtpRequestResult

    /**
     * Submits a request to complete a password-less login to the Salesforce
     * Identity API headless, password-less login flow. This fulfills steps
     * eight, eleven and thirteen of the headless password-less login flow.
     *
     * See https://help.salesforce.com/s/articleView?id=sf.remoteaccess_headless_passwordless_login_public_clients.htm&type=5
     *
     * @param otp A user-entered one-time-password
     * @param otpIdentifier The one-time-password identifier issued by the
     * Salesforce Identity API headless, password-less login flow in the start
     * password-less authorization method.
     * @param otpVerificationMethod The one-time-password verification method
     * used to obtain the OTP identifier
     * @return A native login result indicating success or one of several
     * possible failures, including both in-app and Salesforce Identity API
     * results
     */
    suspend fun submitPasswordlessAuthorizationRequest(
        otp: String,
        otpIdentifier: String,
        otpVerificationMethod: OtpVerificationMethod
    ): NativeLoginResult

    // endregion
}

// region Salesforce Identity API Headless, Password-Less Login Via One-Time-Passcode Data Types

/**
 * An OTP request result.
 * @param nativeLoginResult The overall result of the OTP request
 * @param otpIdentifier On success result, the OTP identifier provided by the
 * API
 */
data class OtpRequestResult(
    val nativeLoginResult: NativeLoginResult,
    val otpIdentifier: String? = null
)

// endregion
// region Common Data Types
/**
 * The possible OTP verification methods.
 * @param identityApiAuthVerificationTypeHeaderValue The expected string value
 * when used in the Salesforce Identity API `Auth-Verification-Type` request
 * header
 */
@Suppress("unused")
enum class OtpVerificationMethod(val identityApiAuthVerificationTypeHeaderValue: String) {
    Email("email"),
    Sms("sms")
}

// endregion

File: flutter_sdk/lib/src/core/auth/native_login_manager.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../rest/http_client_factory.dart';
import 'oauth2.dart';

/// Result of a native login attempt.
enum NativeLoginResult {
  invalidEmail,
  invalidUsername,
  invalidPassword,
  invalidCredentials,
  unknownError,
  success,
}

/// OTP verification method for passwordless and registration flows.
enum OtpVerificationMethod {
  email('email'),
  sms('sms');

  final String identityApiAuthVerificationTypeHeaderValue;
  const OtpVerificationMethod(this.identityApiAuthVerificationTypeHeaderValue);
}

/// Result of an OTP request.
class OtpRequestResult {
  final NativeLoginResult nativeLoginResult;
  final String? otpIdentifier;

  const OtpRequestResult({
    required this.nativeLoginResult,
    this.otpIdentifier,
  });
}

/// Result of starting a registration flow.
class StartRegistrationResult {
  final NativeLoginResult nativeLoginResult;
  final String? email;
  final String? requestIdentifier;

  const StartRegistrationResult({
    required this.nativeLoginResult,
    this.email,
    this.requestIdentifier,
  });
}

/// Manages native credential-based login using Salesforce Identity API headless flows.
///
/// Supports:
/// - Named-user login (username/password via OAuth2 PKCE)
/// - Passwordless OTP login (email/SMS)
/// - User registration (headless)
/// - Password reset (headless)
///
/// Mirrors Android NativeLoginManager for feature parity.
class NativeLoginManager {
  final Uri _loginServer;
  final String _clientId;
  final String _callbackUrl;
  final http.Client _httpClient;

  /// Whether the back button should be shown in native login UI.
  final bool shouldShowBackButton;

  /// Username for biometric re-authentication (if previously stored).
  final String? biometricAuthenticationUsername;

  static final _emailRegex = RegExp(r'^[^@\s]+@[^@\s]+\.[^@\s]+$');
  static const _maxUsernameLength = 80;
  static const _minPasswordLength = 8;
  static final _hasLetterRegex = RegExp(r'[a-zA-Z]');
  static final _hasDigitRegex = RegExp(r'[0-9]');

  NativeLoginManager({
    required Uri loginServer,
    required String clientId,
    required String callbackUrl,
    http.Client? httpClient,
    this.shouldShowBackButton = true,
    this.biometricAuthenticationUsername,
  })  : _loginServer = loginServer,
        _clientId = clientId,
        _callbackUrl = callbackUrl,
        _httpClient = httpClient ?? SalesforceHttpClient();

  /// Performs named-user login with username and password.
  ///
  /// Uses OAuth2 PKCE flow with Basic auth header containing credentials.
  Future<NativeLoginResult> login(String username, String password) async {
    // Validate inputs
    final usernameError = _validateUsername(username);
    if (usernameError != null) return usernameError;

    final passwordError = _validatePassword(password);
    if (passwordError != null) return passwordError;

    try {
      // Generate PKCE
      final codeVerifier = OAuth2.generateCodeVerifier();
      final codeChallenge = OAuth2.generateCodeChallenge(codeVerifier);

      // Encode credentials
      final credentials =
          base64Url.encode(utf8.encode('$username:$password'));

      // Step 1: Authorize with credentials
      final authorizeUrl =
          _loginServer.replace(path: '/services/oauth2/authorize');
      final authResponse = await _httpClient.post(
        authorizeUrl,
        headers: {
          'Authorization': 'Basic $credentials',
          'Content-Type': 'application/x-www-form-urlencoded',
          'Auth-Request-Type': 'Named-User',
        },
        body: {
          'response_type': 'code_credentials',
          'client_id': _clientId,
          'redirect_uri': _callbackUrl,
          'code_challenge': codeChallenge,
        },
      );

      if (authResponse.statusCode != 200) {
        if (authResponse.statusCode == 401 ||
            authResponse.statusCode == 403) {
          return NativeLoginResult.invalidCredentials;
        }
        return NativeLoginResult.unknownError;
      }

      final authJson = jsonDecode(authResponse.body) as Map<String, dynamic>;
      final code = authJson['code'] as String?;
      if (code == null) return NativeLoginResult.unknownError;

      // Step 2: Exchange code for tokens
      await OAuth2.exchangeCode(
        httpClient: _httpClient,
        loginServer: _loginServer,
        clientId: _clientId,
        code: code,
        codeVerifier: codeVerifier,
        callbackUrl: _callbackUrl,
      );

      return NativeLoginResult.success;
    } catch (e) {
      if (e is OAuthFailedException) {
        if (e.tokenErrorResponse?.error == 'invalid_grant') {
          return NativeLoginResult.invalidCredentials;
        }
      }
      return NativeLoginResult.unknownError;
    }
  }

  /// Starts the headless registration flow.
  ///
  /// Sends OTP to the provided email or phone for verification.
  Future<StartRegistrationResult> startRegistration({
    required String email,
    required String firstName,
    required String lastName,
    required String username,
    required String newPassword,
    required String reCaptchaToken,
    required OtpVerificationMethod otpVerificationMethod,
  }) async {
    final usernameError = _validateUsername(username);
    if (usernameError != null) {
      return StartRegistrationResult(nativeLoginResult: usernameError);
    }

    final passwordError = _validatePassword(newPassword);
    if (passwordError != null) {
      return StartRegistrationResult(nativeLoginResult: passwordError);
    }

    if (!_emailRegex.hasMatch(email)) {
      return const StartRegistrationResult(
          nativeLoginResult: NativeLoginResult.invalidEmail);
    }

    try {
      final url = _loginServer.replace(
          path: '/services/auth/headless/init/registration');
      final response = await _httpClient.post(
        url,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Auth-Request-Type': 'Named-User',
          'Auth-Verification-Type':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
        body: {
          'email': email,
          'first_name': firstName,
          'last_name': lastName,
          'username': username,
          'newpassword': newPassword,
          'recaptcha': reCaptchaToken,
          'verificationmethod':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
      );

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return StartRegistrationResult(
          nativeLoginResult: NativeLoginResult.success,
          email: json['email'] as String?,
          requestIdentifier: json['identifier'] as String?,
        );
      }
      return const StartRegistrationResult(
          nativeLoginResult: NativeLoginResult.unknownError);
    } catch (_) {
      return const StartRegistrationResult(
          nativeLoginResult: NativeLoginResult.unknownError);
    }
  }

  /// Completes the registration flow with the OTP received.
  Future<NativeLoginResult> completeRegistration({
    required String otp,
    required String requestIdentifier,
    required OtpVerificationMethod otpVerificationMethod,
  }) async {
    try {
      final codeVerifier = OAuth2.generateCodeVerifier();
      final codeChallenge = OAuth2.generateCodeChallenge(codeVerifier);

      final url =
          _loginServer.replace(path: '/services/oauth2/authorize');
      final response = await _httpClient.post(
        url,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Auth-Request-Type': 'Named-User',
          'Auth-Verification-Type':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
        body: {
          'response_type': 'code_credentials',
          'client_id': _clientId,
          'redirect_uri': _callbackUrl,
          'code_challenge': codeChallenge,
          'otp': otp,
          'identifier': requestIdentifier,
          'verificationmethod':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
      );

      if (response.statusCode != 200) {
        return NativeLoginResult.unknownError;
      }

      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final code = json['code'] as String?;
      if (code == null) return NativeLoginResult.unknownError;

      await OAuth2.exchangeCode(
        httpClient: _httpClient,
        loginServer: _loginServer,
        clientId: _clientId,
        code: code,
        codeVerifier: codeVerifier,
        callbackUrl: _callbackUrl,
      );

      return NativeLoginResult.success;
    } catch (_) {
      return NativeLoginResult.unknownError;
    }
  }

  /// Starts the password reset flow.
  Future<NativeLoginResult> startPasswordReset({
    required String username,
    required String reCaptchaToken,
  }) async {
    final usernameError = _validateUsername(username);
    if (usernameError != null) return usernameError;

    try {
      final url = _loginServer.replace(
          path: '/services/auth/headless/forgot_password');
      final response = await _httpClient.post(
        url,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: {
          'username': username,
          'recaptcha': reCaptchaToken,
        },
      );

      return response.statusCode == 200
          ? NativeLoginResult.success
          : NativeLoginResult.unknownError;
    } catch (_) {
      return NativeLoginResult.unknownError;
    }
  }

  /// Completes the password reset with the OTP and new password.
  Future<NativeLoginResult> completePasswordReset({
    required String username,
    required String otp,
    required String newPassword,
  }) async {
    final passwordError = _validatePassword(newPassword);
    if (passwordError != null) return passwordError;

    try {
      final url = _loginServer.replace(
          path: '/services/auth/headless/forgot_password/reset');
      final response = await _httpClient.post(
        url,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: {
          'username': username,
          'otp': otp,
          'newpassword': newPassword,
        },
      );

      return response.statusCode == 200
          ? NativeLoginResult.success
          : NativeLoginResult.unknownError;
    } catch (_) {
      return NativeLoginResult.unknownError;
    }
  }

  /// Submits an OTP request for passwordless login.
  Future<OtpRequestResult> submitOtpRequest({
    required String username,
    required String reCaptchaToken,
    required OtpVerificationMethod otpVerificationMethod,
  }) async {
    final usernameError = _validateUsername(username);
    if (usernameError != null) {
      return OtpRequestResult(nativeLoginResult: usernameError);
    }

    try {
      final url = _loginServer.replace(
          path: '/services/auth/headless/init/passwordless/login');
      final response = await _httpClient.post(
        url,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Auth-Verification-Type':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
        body: {
          'username': username,
          'recaptcha': reCaptchaToken,
          'verificationmethod':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
      );

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return OtpRequestResult(
          nativeLoginResult: NativeLoginResult.success,
          otpIdentifier: json['identifier'] as String?,
        );
      }
      return const OtpRequestResult(
          nativeLoginResult: NativeLoginResult.unknownError);
    } catch (_) {
      return const OtpRequestResult(
          nativeLoginResult: NativeLoginResult.unknownError);
    }
  }

  /// Submits the OTP for passwordless authorization.
  Future<NativeLoginResult> submitPasswordlessAuthorizationRequest({
    required String otp,
    required String otpIdentifier,
    required OtpVerificationMethod otpVerificationMethod,
  }) async {
    try {
      final codeVerifier = OAuth2.generateCodeVerifier();
      final codeChallenge = OAuth2.generateCodeChallenge(codeVerifier);

      final url =
          _loginServer.replace(path: '/services/oauth2/authorize');
      final response = await _httpClient.post(
        url,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Auth-Request-Type': 'Passwordless-Login',
          'Auth-Verification-Type':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
        body: {
          'response_type': 'code_credentials',
          'client_id': _clientId,
          'redirect_uri': _callbackUrl,
          'code_challenge': codeChallenge,
          'otp': otp,
          'identifier': otpIdentifier,
          'verificationmethod':
              otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
        },
      );

      if (response.statusCode != 200) {
        return NativeLoginResult.unknownError;
      }

      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final code = json['code'] as String?;
      if (code == null) return NativeLoginResult.unknownError;

      await OAuth2.exchangeCode(
        httpClient: _httpClient,
        loginServer: _loginServer,
        clientId: _clientId,
        code: code,
        codeVerifier: codeVerifier,
        callbackUrl: _callbackUrl,
      );

      return NativeLoginResult.success;
    } catch (_) {
      return NativeLoginResult.unknownError;
    }
  }

  NativeLoginResult? _validateUsername(String username) {
    if (username.isEmpty || username.length > _maxUsernameLength) {
      return NativeLoginResult.invalidUsername;
    }
    if (!_emailRegex.hasMatch(username)) {
      return NativeLoginResult.invalidEmail;
    }
    return null;
  }

  NativeLoginResult? _validatePassword(String password) {
    if (password.length < _minPasswordLength) {
      return NativeLoginResult.invalidPassword;
    }
    if (!_hasLetterRegex.hasMatch(password) ||
        !_hasDigitRegex.hasMatch(password)) {
      return NativeLoginResult.invalidPassword;
    }
    return null;
  }
}
